Naming conventions rules (#21)
BlueBrain/hpc-coding-conventions: cpp/NamingConventions.md

# BlueBrain HPC Team C++ Naming Conventions

This document describes how C++ symbols should be named.

## Common Conventions

### Use descriptive names

```cpp
// no
int bufsz;

// yes
int buffer_size;
```

Generally use `snake_case` style as it is compliant with spell checkers.

#### Exception

> Function parameters and local variables used only within less than 3 lines can break this rule, for instance:

Reviewer: ```suggestion
Function parameters and local variables used only within the next 3 lines can break this rule, for instance:
```
I find `within less than` confusing, since `within` already containing a `less than the limit` sort of. I would personally raise that limit a little, too, but oh well.

```cpp
struct Dimension {
    const int width;
    const int height;

    Dimension(int w, int h)
        : width(w)
	, height(h) {}
};
```

### Functions and variables start with a lower case

```cpp
void my_function() {
	int my_var;
};
```

### Types names use camel case

```cpp
class MyClass
using MyClassVector = std::vector<MyClass>
```

### Template parameter names use camel case: `InputIterator`

```cpp
template <class InputIterator, class Distance>
void advance(InputIterator& it, Distance n);
```

### Constants are all upper case with underscores

```cpp
const double AVOGADRO = 6.022140857e23;
const int LONG_MAX = 9223372036854775807L;
```

> ## Distinguish symbols

Reviewer 2: This practice i really dislike. I find it makes the code look very ugly. I don't really have an alternative, since I would simply not do it

Distinguish private member variables, function parameters, and local variables.
There are several alternatives we can consider:

### Add a prefix to both private members and function parameters

* unexposed member variables are prefixed with `m_`

> * function parameters are prefixed with `t_`

Reviewer: I still find this confusing with the `_t` for `type` stuff, and would actually discourage it.

* do not use prefix for local variables and exposed member variables

#### Examples

```cpp
struct Size {
    int width;
    int height;

    Size(int t_width, int t_height)
        : width(t_width)
        , height(t_height) {}
};

class PrivateSize {
  public:
    int width() const {
        return m_width;
    }
    int height() const {
        return m_height;
    }
    PrivateSize(int t_width, int t_height)
        : m_width(t_width)
        , m_height(t_height) {}

  private:
    int m_width;
    int m_height;
};
```

```cpp
class MyClass {
  public:
    MyClass(int t_data)
        : m_data(t_data) {}

    int data() const {
        return m_data;
    }

  private:
    int m_data;
};
```

### Add prefix to private members and a suffix to function parameters

> * unexposed member variables are prefixed with `m_`

Reviewer 2: maybe I would leave function arguments alone but prefix unexposed member vars with `_`.

Author: I take note that you consider marking unexposed members. But for reasons I explained previously in this discussion, prefixing with `_` is not an option.

Reviewer 3: I don't have better suggestion but `_` and prefix/suffix stuff doesn't look that nice.

> but prefix unexposed member vars with `_`  (Or, suffix/suffix)

Only some member variables with `m_` and others without : I would prefer declare them uniformly.

Author: > Only some member variables with `m_` and others without : I would prefer use them uniformly.

The "others" are the public class member variables, which are directly used so IMHO you don't want to prefix/suffix them.

> * function parameters are suffixed with `_`

Reviewer 2: that's the only exception that I could *maybe* live with.. since the alternative (using short names or hunting around for other acceptable variable names) would be even worse

Reviewer: I think it's good for collisions, as in constructors, and could be dropped if it is unambiguous. I'm not really a fan of too much pre/suffixing. We could also prefix *all* local variables with `l_`…

* do not use prefix for local variables and exposed member variables

#### Examples

```cpp
struct Size {
    int width;
    int height;

    Size(int width_, int height_)
        : width(width_)
        , height(height_) {}
};

class PrivateSize {
  public:
    int width() const {
        return m_width;
    }
    int height() const {
        return m_height;
    }
    PrivateSize(int width_, int height_)
        : m_width(width_)
        , m_height(height_) {}

  private:
    int m_width;
    int m_height;
};
```

```cpp
class MyClass {
  public:
    MyClass(int data_)
        : m_data(data_) {}

    int data() const {
        return m_data;
    }

  private:
    int m_data;
};
```

## Comments

> Comment blocks should use `//`, not `/* */`. Using `//` makes it much easier to comment

Reviewer 2: on one hand I agree.. on the other hand I find docstring formatting in the `/* */` comments much neater

Reviewer 3: agree...for oner liner `//` seems ok but anything larger I have been preferring /* */

Reviewer 4: I agree that docstrings look nicer with `/**   */`
But in the middle of a function I agree with the argument of @tristan0x.
Anyway, to comment out a large region one can always do `#if 0`  ... `#endif`

# C++ Naming Conventions

This document provides a set of recommendations regarding how things should be named
in C++ projects.

Because there is no automatic way to ensure that a code change follows this guideline,
it is the responsibility of the code reviewers to enforce these rules over time.

Should a project decide to not apply some or all of the recommendations
below, it may provide written guidelines to the contributors with the rules
that should be followed instead, in a `CONTRIBUTING.md` document for instance.

## File Names

Follow [File Names](https://google.github.io/styleguide/cppguide.html#General_Naming_Rules)
convention of Google C++ Style Guide but use the following extension instead:

* C++ files: `.cpp`
* C++ header: `.hpp`
* C++ template implementations: `.ipp`

## Use descriptive variable names

```cpp
// no
int bufsz;

// yes

> int buffer_size;

Reviewer 2: one thing I come across every once in a while.. what do you do when the variable name would be really very long if you'd write it out without abbreviating?

Author: This is not a strict rule banning abbreviation. For instance, in the URL mentioned below:

   ```cpp
   int lstm_size;             // "LSTM" is a common machine learning abbreviation.
   ```

The idea is to be aware of the consequences of losing meaning just to save a few characters.

But I think the `buffer_size` example doesn't entirely highlight the meaning of this section. The point here is mainly to banish variables named like `count` for instance. I will make some updates. Thank you

Author: Hello @ohm314 
I tried to address your question in 5f3377b2
Please, _Resolve conversation_ or comment.

Reviewer 2: Sorry @tristan0x I completely missed that. Your comment makes sense and I agree

```

See [General Naming Rules](https://google.github.io/styleguide/cppguide.html#General_Naming_Rules)
in Google C++ Style Guide.

Generally use `snake_case` style as it is compliant with spell checkers.

### Exception

Function parameters and local variables used only within less than 3 lines can break this rule, for instance:

```cpp
struct Dimension {
    const int width;
    const int height;

    Dimension(int w, int h)
        : width(w)
	, height(h) {}
};
```

## Functions and variables start with a lower case

```cpp
void my_function() {
	int my_var;
};
```

## Types names use camel case

```cpp
class MyClass
using MyClassVector = std::vector<MyClass>
```

## Template parameter names use camel case: `InputIterator`

```cpp
template <class InputIterator, class Distance>
void advance(InputIterator& it, Distance n);
```

## Constants are all upper case with underscores

```cpp
const double AVOGADRO = 6.022140857e23;

> const int LONG_MAX = 9223372036854775807L;

Reviewer: Shouldn't this come from limits?

Author: `LONG_MAX` is maybe not the best example.

Author: I changed the example to `LIGHT_SPEED`.
